test/support/warning_extractor.rb

```diff
@@ -0,0 +1,51 @@
+require 'set'
+
+module RegexpParserTest
+  class Warning
+    class Filter
+      def initialize(whitelist)
+        @whitelist = whitelist
+      end
+
+      def assert_expected_warnings_only
+        original = $stderr
+        $stderr  = Extractor.new(original, @whitelist)
+
+        yield
+
+        assert_no_warnings($stderr.warnings)
+      ensure
+        $stderr = original
+      end
+
+    private
+
+      def assert_no_warnings(warnings)
+        fail "Unexpected warnings: #{warnings.to_a.join("\n")}" if warnings.any?
```

Comment:
@mbj I ended up actually using the `%s` change but I also just extracted the message to a constant within a dedicated errors class